src/NLog/Internal/PropertySetter.cs

```diff
@@ -60,8 +75,50 @@ private PropertySetter([NotNull] PropertyInfo property)
         /// <param name="bindingFlags"></param>
         /// <returns></returns>
         [CanBeNull]
-        public static PropertySetter<T> TryCreate([NotNull] T obj, [NotNull] string propertyName,
+        public static PropertySetter<T> TryCreate<T>([NotNull] T obj, [NotNull] string propertyName,
             BindingFlags bindingFlags = BindingFlags.Public | BindingFlags.Instance | BindingFlags.IgnoreCase) // note: BindingFlags.Default is not available for .NET Standard 1.3
+        {
+            if (PropertySetter<T>.TryGetProperty(obj, propertyName, bindingFlags, out var property))
+            {
+                return null;
+            }
+
+            return new PropertySetter<T>(property);
+        }
+    }
+
+    /// <summary>
+    /// Compiled version for setting a property value.
+    /// </summary>
+    /// <typeparam name="T">The type of the property. Not used for detection, but only for compile-time safe <see cref="SetValue"/></typeparam>
+    internal class PropertySetter<T>
+    {
+        private readonly ReflectionHelpers.LateBoundMethod _propertySetter;
+
+        /// <summary>
+        /// array with 1 item. The property will always have one property
+        /// </summary>
+        private readonly object[] _arrayWith1Item = new object[1];
```

Comment:
Having the object-reuse here will give race-condition issues for those not aware. I would move this object-reuse out of the PropertySetter-class (Have SetValue-method that takes array)

Reply:
The array-reuse is mostly important for custom DbType. For Connection-properties then a single-item-array allocation is nothing compare to the allocation performed by DbCommand.

Reply:
I think we should fix the array thing outside this PR. I think the latebound creator could be improved. Will create a new PR for that.